chain/mining.go

```diff
@@ -75,27 +76,35 @@ func (bm *BuiltinMining) BuildBlock(ctx context.Context, height uint32, chordID
 		default:
 		}
 
+		if txCount >= int(config.Parameters.NumTxnPerBlock) {
+			break
+		}
+
 		txn := txnCollection.Peek()
 		if txn == nil {
 			break
 		}
 
-		if txn.UnsignedTx.Fee < int64(config.Parameters.MinTxnFee) {
-			log.Warning("transaction fee is too low")
-			txnCollection.Pop()
-			continue
+		if config.Parameters.MinTxnFee > 0 {
+			if txn.UnsignedTx.Fee == 0 {
+				if uint32(zeroFeeTxCount) >= config.Parameters.NumZeroFeeTxnPerBlock {
+					log.Warning("too many zero fee transactions")
+					break
+				}
+			} else {
+				if txn.UnsignedTx.Fee < int64(config.Parameters.MinTxnFee) {
+					log.Warning("transaction fee is too low")
+					txnCollection.Pop()
```

Comment:
It should not collection.Pop() but should zeroFeeTxCount++ if zeroFeeTxCount not reach NumZeroFeeTxnPerBlock.

Reply:
It pop() because the txn will never be packed into the block by this miner.
zeroFeeTxCount should only increase if and only if a zero fee txn is packed into the block.
PS: this is the old logic and has not been changed

Reply:
It is unreasonable a `non-zero but less MinTxnFee fee` was rejected by peek `a zero fee` preferentially.

Reply:
`Pop()` has to be called whenever `continue` is called, otherwise it will get stuck

Reply:
```
Fee > MinTxnFee   // Accepted
Fee < MinTxnFee   // Rejected
Fee == 0          // Accepted * NumZeroFeeTxnPerBlock
```
It is unfair for `Fee < MinTxnFee`